File: chess.py
import copy
import math
from typing import *

dy = [-1, 1, -2, 2, -1, 1, -2, 2]
dx = [-2, -2, -1, -1, 2, 2, 1, 1]

infinity = math.inf


class Game:
    """문제를 정의하기 위한 Problem 클래스에 대응되는 게임 정의를 위한 클래스.
    경로 비용과 목표 검사 대신 각 상태에 대한 효용 함수와 종료 검사로 구성됨.
    게임을 정의하려면 이 클래스의 서브 클래스를 만들어서
    actions, result, is_terminal, utility를 구현하면 됨.
    필요에 따라 게임의 초기 상태를 지정하려면,
    클래스 생성자에서 초기 상태를 initial 에 세팅하면 됨."""

    def actions(self, state):
        """주어진 상태에서 허용 가능한 수(move) 리스트"""
        raise NotImplementedError

    def result(self, state, move):
        """주어진 상태(state)에서 수(move)를 두었을 때의 결과 상태 리턴"""
        raise NotImplementedError

    def is_terminal(self, state):
        """state가 종료 상태이면 True 리턴"""
        return not self.actions(state)

    def utility(self, state, player):
        """종료 상태 state에서 게임이 종료됐을 때 player의 효용 함수 값"""
        raise NotImplementedError


class Node:
    """
    상태공간 노드입니다.
    """

    def __init__(self, size: int, y: int, x: int, board: List[List], point: int, turn: int):
        self.size: int = size
        self.y: int = y
        self.x: int = x
        self.board: List[List] = board
        self.point: int = point
        self.turn: int = turn  # 현재 말의 턴을 저장합니다. 0: max, 1: min

    def utility(self):
        return self.point


class ChessGame(Game):
    def __init__(self):
        self.initial = Node()

    def actions(self, state) -> List[Tuple[int, int]]:
        result = []

        for i in range(8):
            ny, nx = state.y + dy[i], state.x + dx[i]

            if not (0 <= ny < state.size):
                continue
            if not (0 <= nx < state.size):
                continue
            if state.board[ny][nx] == -1:
                continue

            result.append((ny, nx))

        return result

    def result(self, state: Node, move: Tuple[int, int]):
        next_state = copy.deepcopy(state)
        y, x = move
        next_state.point += next_state.board[y][x]
        next_state.board[y][x] = -1
        return next_state

    def utility(self, state: Node, player: int):
        return state.point


def alphabeta_search(game, state):
    """알파-베타 가지치기를 사용하여 최고의 수를 결정하기 위한 게임 트리 탐색."""

    player = state.to_move

    def max_value(state, alpha, beta):
        if game.is_terminal(state):
            return game.utility(state, player), None
        v, move = -infinity, None
        for a in game.actions(state):
            v2, _ = min_value(game.result(state, a), alpha, beta)
            if v2 > v:
                v, move = v2, a
                alpha = max(alpha, v)
            if v >= beta:
                return v, move
        return v, move

    def min_value(state, alpha, beta):
        if game.is_terminal(state):
            return game.utility(state, player), None
        v, move = +infinity, None
        for a in game.actions(state):
            v2, _ = max_value(game.result(state, a), alpha, beta)
            if v2 < v:
                v, move = v2, a
                beta = min(beta, v)
            if v <= alpha:
                return v, move
        return v, move

    return max_value(state, -infinity, +infinity)


def play_game(game, strategies: dict, verbose=False) -> Node:
    """번갈아 가면서 두는 게임 진행.
    strategies: {참가자 이름: 함수} 형태의 딕셔너리.
    함수(game, state)는 상태 state에서 참가자의 수를 찾는 함수"""
    state = game.initial
    while not game.is_terminal(state):
        player = state.to_move
        move = strategies[player](game, state)
        state = game.result(state, move)
        if verbose:
            print('Player', player, 'move:', move)
            print(state)
    return state


def game_player(search_algorithm):
    """지정된 탐색 알고리즘을 사용하는 플레이어: (game, state)를 입력 받아 move를 리턴하는 함수."""
    return lambda game, state: search_algorithm(game, state)[1]


if __name__ == '__main__':
    n: int = int(input("체스판의 크기를 입력해주세요: "))

    print("n * n 체스판에 넣을 점수를 입력해주세요")
    board: List[List[int]] = [list(map(int, input(f"{i}번째 줄: "))) for i in range(n)]

    print(
        f"플레이어 A와 B의 최대 점수차이는 {play_game(ChessGame(), {'a': game_player(alphabeta_search), 'b': game_player(alphabeta_search)}).utility()}점 입니다.")

File: chess.cpp
#include <iostream>
#include <vector>

using namespace std;

typedef pair<int, int> ii;
typedef pair<int, ii> iii;

// 상태공간에 기반한 게임 클래스입니다.
class Game;

// 상태공간에서의 각 상태를 나타내는 노드 클래스입니다.
class Node;

// 하나의 나이트를 가지고 땅따먹기를 하는 체스게임의 축소 문제입니다.
class ChessGame;

// 알파-베타 탐색을 수행하는 함수입니다.
int AlphaBetaSearch(Game *game, Node *state);

// 알파-베타 탐색의 MaxValue 부분입니다.
iii MaxValue(Game *game, Node *state, int alpha, int beta);

// 알파-베타 탐색의 MinValue 부분입니다.
iii MinValue(Game *game, Node *state, int alpha, int beta);

// 나이트의 다음 이동 경로를 저장하는 변수입니다.
const int dy[8] = {-2, -2, -1, -1, 2, 2, 1, 1};
const int dx[8] = {-1, 1, -2, 2, -1, 1, -2, 2};

class Game
{
public:
	virtual void getInitial(Node *out) = 0;

	// 주어진 상태에서 허용 가능한 수 리스트를 반환
	virtual void actions(Node *state, vector<ii> *result) = 0;

	// 주어진 상태에서 수를 두었을 때의 결과 상태 반환
	virtual void result(Node *state, ii *move, int turn, Node *result) = 0;

	// 종료 상태 여부를 반환
	bool isTerminal(Node *state)
	{
		vector<ii> action_list;
		actions(state, &action_list);
		return action_list.empty();
	}

	// 종료 상태에서 게임이 종료되었을 때의 효용값 반환
	virtual int utility(Node *state) = 0;
};

class Node
{
public:
	int size;
	int y;
	int x;
	vector<vector<int> > board;
	int point;

	Node(int size, int y, int x, vector<vector<int> > *board)
	{
		this->size = size;
		this->y = y;
		this->x = x;
		this->board.assign(board->begin(), board->end());
		this->point = 0;
	}

	explicit Node(Node *node)
	{
		size = node->size;
		y = node->y;
		x = node->x;
		board.assign(node->board.begin(), node->board.end());
		point = node->point;
	}

	Node(Node const &node) : size(node.size), y(node.y), x(node.x), point(node.point)
	{
		board.assign(node.board.begin(), node.board.end());
	}

	Node()
	{
		size = 0;
		y = 0;
		x = 0;
		point = 0;
	}

	void init(Node *node)
	{
		size = node->size;
		y = node->y;
		x = node->x;
		board.assign(node->board.begin(), node->board.end());
		point = node->point;
	}

	// 현 상태의 효용값 반환
	int utility() const
	{
		return point;
	}

	// 주어진 좌표에 말을 놓을 수 있는지 체크
	bool isValidPos(int y, int x)
	{
		if (y < 0 || size <= y)
		{
			return false;
		}
		if (x < 0 || size <= x)
		{
			return false;
		}
		if (board[y][x] == -1)
		{
			return false;
		}
		return true;
	}
};

class ChessGame : public Game
{
public:
	Node initial;

	ChessGame(int size, int y, int x, vector<vector<int> > *board) : initial(size, y, x, board)
	{

	}

	explicit ChessGame(ChessGame *other) : initial(&(other->initial))
	{

	}

	void getInitial(Node *out)
	{
		out = &initial;
	}

	void actions(Node *state, vector<ii> *result)
	{
		result->clear();

		// 8개의 이동 경로를 순회하면서 가능한 이동 경로를 저장합니다.
		for (int i = 0; i < 8; i++)
		{
			int ny = state->y + dy[i];
			int nx = state->x + dx[i];

			if (!state->isValidPos(ny, nx))
			{
				continue;
			}

			result->push_back(make_pair(ny, nx));
		}
	}

	void result(Node *state, ii *move, int turn, Node *result)
	{
		Node nextState(state);
		int y = move->first;
		int x = move->second;

		// turn이 0인 경우 Max, 1인 경우 Min을 의미합니다.
		if (turn == 0)
		{
			nextState.point += nextState.board[y][x];
		} else
		{
			nextState.point -= nextState.board[y][x];
		}
		nextState.board[y][x] = -1;

		// 새로운 상태를 result 변수에 복사합니다.
		result->init(&nextState);
	}

	int utility(Node *state)
	{
		return state->utility();
	}
};

int AlphaBetaSearch(Game *game, Node *state)
{
	// Max가 게임을 시작하는 것으로 알파-베타 탐색을 시작합니다.
	return MaxValue(game, state, -1e9, 1e9).first;
}

iii MaxValue(Game *game, Node *state, int alpha, int beta)
{
	if (game->isTerminal(state))
	{
		return make_pair(game->utility(state), ii());
	}

	int v = -1e9;
	ii move = ii();

	vector<ii> actions;
	game->actions(state, &actions);

	for (int i = 0; i < actions.size(); i++)
	{
		Node result;
		game->result(state, &actions[i], 0, &result);
		int v2 = MinValue(game, &result, alpha, beta).first;

		if (v2 > v)
		{
			v = v2;
			move = actions[i];
			alpha = max(alpha, v);
		}

		if (v >= beta)
		{
			return make_pair(v, move);
		}
	}

	return make_pair(v, move);
}

iii MinValue(Game *game, Node *state, int alpha, int beta)
{
	if (game->isTerminal(state))
	{
		return make_pair(game->utility(state), ii());
	}

	int v = 1e9;
	ii move = ii();

	vector<ii> actions;
	game->actions(state, &actions);

	for (int i = 0; i < actions.size(); i++)
	{
		Node result;
		game->result(state, &actions[i], 1, &result);
		int v2 = MaxValue(game, &result, alpha, beta).first;

		if (v2 < v)
		{
			v = v2;
			move = actions[i];
			beta = min(beta, v);
		}

		if (v <= alpha)
		{
			return make_pair(v, move);
		}
	}

	return make_pair(v, move);
}

int main()
{
	int size = 0;

	cout << "체스판의 크기를 입력해주세요: ";

	cin >> size;
	cin.ignore();

	vector<vector<int> > board(size, vector<int>(size));

	cout << "체스판에 넣을 점수를 입력해주세요" << endl;

	for (int i = 0; i < size; i++)
	{
		for (int j = 0; j < size; j++)
		{
			cin >> board[i][j];
			cin.ignore();
		}
	}

	ChessGame game(size, 0, 0, &board);

	int ans = AlphaBetaSearch(&game, &game.initial);

	cout << "플레이어 A와 B의 최대 점수차이는 " << ans << " 입니다." << endl;
}